src/routes/api/client-credentials/microsoft/token.ts

```diff
@@ -0,0 +1,163 @@
+import { webexHttpPeopleResource } from '$lib/webex/http-wrapper';
+import { MikroORM } from '@mikro-orm/core';
+import { Activation } from '../../../../database/entities';
+import config from '../../../../../mikro-orm.config';
+import humps from 'humps';
+import env from '$lib/environment';
+import * as jose from 'jose';
+import { base } from '$app/paths';
+
+/** @typedef {import('class-validator').ValidationError} ValidationError */
+
+const onFailure = async (e: Response | Error) => {
+  if (e instanceof Response && e.status !== 401) {
+    return e.status === 428
+      ? { status: e.status, body: await e.json(), headers: { 'Skip-Reporting': true } }
+      : { status: e.status, body: await e.json() };
+  } else {
+    console.log(e);
+    return { status: 500 };
+  }
+};
+
+/**
+ * @param {RequestEvent} requestEvent
+ *
+ * @returns {Promise<
+ *   | { body: { query: ValidationError[] }; status: 400 }
+ *   | { body: ResponseDTO; status: 200 }
+ *   | { body: any; status: number }
+ *   | { status: 500 }
+ * >}
+ */
+export const POST = async (requestEvent: RequestEvent) => {
+  class RequestQueryDTO {
+    @Expose()
+    @IsNotEmpty()
+    @IsUUID(4)
+    public readonly activationId!: string;
+  }
+
+  class RequestHeaderDTO {
+    @Expose()
+    @IsNotEmpty()
+    @Transform(({ obj }) => obj?.authorization?.toString()?.replace(/^Bearer /i, ''))
+    public readonly authorization!: string;
+  }
+
+  class ResponseDTO implements ToJSON {
+    @Expose()
+    @Transform(({ obj, value }) => value ?? obj.expires_in, { toClassOnly: true })
+    public readonly expiresIn!: number;
+
+    @Expose()
+    @Transform(({ value }) => new Date(value).toUTCString(), { toPlainOnly: true })
+    public readonly expiresAt?: string;
+
+    @Expose()
+    @Transform(({ obj, value }) => value ?? obj.token_type, { toClassOnly: true })
+    public readonly tokenType!: string;
+
+    @Expose()
+    @Transform(({ obj, value }) => value ?? obj.access_token, { toClassOnly: true })
+    public readonly accessToken!: string;
+
+    toJSON(): Exclude<JSONValue, ToJSON> {
+      return instanceToPlain(this);
+    }
+  }
+
+  // validate request query
+  const searchParams = Object.fromEntries(requestEvent.url.searchParams);
+  const query = plainToInstance(RequestQueryDTO, searchParams, classTransformOptions);
+  const queryValidationErrors = validateSync(query, classValidationOptions);
+  if (queryValidationErrors.length > 0) {
+    return { status: 400, body: { query: queryValidationErrors } };
+  }
+
+  // validate request header
+  const obj = humps.camelizeKeys(Object.fromEntries(requestEvent.request.headers.entries()));
+  const header = plainToInstance(RequestHeaderDTO, obj, classTransformOptions);
+  const headerValidationErrors = validateSync(header, classValidationOptions);
+  if (headerValidationErrors.length > 0) {
+    return { status: 401 };
+  }
+
+  const orm = await MikroORM.init({ ...config, ...{ entities: [Activation] } });
```

Comment:
Can't you just access the entity attached to the request Event local property instead of initiating another Driver?

Reply:
Good point. But `/api` routes are not attached to a session and also don't have access to `event.locals.db`. The API routes that are dependent on the DB are very limited (3 at present, including this one). Even with that, those 3 need access to at most 1 or 2 tables (entities). Hence initialization is done on case by case basis.